{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns\n",
    "from matplotlib.colors import LinearSegmentedColormap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data of each sample\n",
    "data = dict()\n",
    "\n",
    "data['HCH3944'] = {\n",
    "    'sampling_freq': 7,\n",
    "    'first_lineage': {\n",
    "        'name': 'BA.5.2*',\n",
    "        'sig': ['C1627T', 'G12160A', 'T21765-', 'A21766-', 'C21767-', 'A21768-', 'T21769-', 'G21770-', 'T22917G', 'T23018G', 'G26529A', 'C27889T', 'A28330G']\n",
    "    },\n",
    "    'second_lineage': {\n",
    "        'name': 'BA.2.3.7',\n",
    "        'sig': ['C2156T', 'C8991T', 'C9866T', 'A21851G', 'A23040G', 'G25314T', 'C25810T', 'C26858T', 'A27259C', 'G27382C', 'A27383T', 'T27384C', 'G29239A']\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "data['HCH3987'] = {\n",
    "    'sampling_freq': 7,\n",
    "    'first_lineage': {\n",
    "        'name': 'BA.5.2.1*',\n",
    "        'sig': ['C1627T', 'G12160A', 'T21765-', 'A21766-', 'C21767-', 'A21768-', 'T21769-', 'G21770-', 'T22917G', 'T23018G', 'G26529A', 'A27038G', 'C27889T', 'A28330G']\n",
    "    },\n",
    "    'second_lineage': {\n",
    "        'name': 'BA.2.3.7',\n",
    "        'sig': ['C2156T', 'C8991T', 'C9866T', 'A21851G', 'A23040G', 'G25314T', 'C25810T', 'C26858T', 'A27259C', 'G27382C', 'A27383T', 'T27384C', 'G29239A']\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "data['HCH4149'] = {\n",
    "    'sampling_freq': 7,\n",
    "    'first_lineage': {\n",
    "        'name': 'BA.2.75.5*',\n",
    "        'sig': ['C3796T', 'C3927T', 'C4586T', 'C5183T', 'A12444G', 'G15451A', 'C20235T', 'A22001G', 'T22016C', 'C22033A', 'A22190G',\\\n",
    "                'G22331A', 'G22577C', 'G22599C', 'A22629C', 'G22898A', 'T22942G', 'T23031C', 'C25416T', 'A26275G']\n",
    "    },\n",
    "    'second_lineage': {\n",
    "        'name': 'BA.2.3.7',\n",
    "        'sig': ['C2156T', 'C8991T', 'G29239A', 'C25810T', 'G25314T', 'A21851G', 'A23040G']\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "data['HCH4333'] = {\n",
    "    'sampling_freq': 5,\n",
    "    'first_lineage': {\n",
    "        'name': 'BA.5.3.1*',\n",
    "        'sig': ['C1931A', 'T2954C', 'C11750T', 'G12160A', 'T14257C', 'G16935A', 'T21765-', 'A21766-', 'C21767-', 'A21768-', 'T21769-', 'G21770-', 'A22893C', \\\n",
    "                'T22917G', 'T22942A', 'T23018G', 'G26529A', 'C27889T', 'C28312T', 'G28681T']\n",
    "    },\n",
    "    'second_lineage': {\n",
    "        'name': 'BA.2.3.7',\n",
    "        'sig': ['C2156T', 'C8991T', 'C9866T', 'C26858T', 'T27384C', 'G27382C', 'A27259C', 'C25810T', 'G29239A', 'G25314T', 'A21851G', 'A27383T', 'A23040G']\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ======================  VARIABLES TO BE ASSIGNED  ===================================\n",
    "patient_id = 'HCH4149'\n",
    "\n",
    "fig_size = (18, 7.5) # Figure size (19, 6.2) or (18, 7.5) \n",
    "subtitle1 = 'A'\n",
    "subtitle2 = 'B'\n",
    "\n",
    "# ========= Modify the filenames accordingly ========\n",
    "working_directory_path = '.'\n",
    "sample_info_file = 'NHRI_Run9-17_sample information.xlsx'\n",
    "# modify the list of column names if other information in the info_file to be annotated on the plot\n",
    "columns_for_info_label = ['Collection date', 'Sequence coverage', 'Pango lineage'] \n",
    "total_reads_fname = 'alignedReads.txt'\n",
    "base_freq_fname = 'base.fqy'\n",
    "SARS2ref = 'NC_045512.2' # SARS-CoV-2 reference ID\n",
    "# ========================================================================================\n",
    "\n",
    "sampling_freq = data[patient_id]['sampling_freq']   # number of tests taken\n",
    "first_lineage_sig = data[patient_id]['first_lineage']['sig']\n",
    "second_lineage_sig = data[patient_id]['second_lineage']['sig']\n",
    "\n",
    "# sample information list\n",
    "df_sample_info = pd.read_excel(f'{working_directory_path}/data/{sample_info_file}')\n",
    "# column value formatting, comment out if not needed\n",
    "df_sample_info['Sequence coverage'] = df_sample_info['Sequence coverage'].apply(lambda x: str(x)+'%')\n",
    "df_sample_info['Collection date'] = df_sample_info['Collection date'].apply(lambda x: x.strftime(\"%Y-%m-%d\"))\n",
    "\n",
    "\n",
    "# sorting the signatures by their position\n",
    "first_lineage_sig = sorted(first_lineage_sig, key=lambda x: int(x[1:-1]))\n",
    "second_lineage_sig = sorted(second_lineage_sig, key=lambda x: int(x[1:-1]))\n",
    "\n",
    "# list of all sample No. for this patient\n",
    "sample_no = [patient_id] + [f'{patient_id}-{str(i)}' for i in range(2,sampling_freq+1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a dictionary to hold data of each sample\n",
    "dict_samples = dict()\n",
    "\n",
    "# coordinates of the signatures\n",
    "coordinates = [int(c[1:-1]) for c in first_lineage_sig + second_lineage_sig]\n",
    "\n",
    "\n",
    "# Iterate through each samples of this patient\n",
    "for sample in sample_no: \n",
    "    # Sample directory\n",
    "    sample_directory_path = f'{working_directory_path}/data/{patient_id}/{sample}'\n",
    "\n",
    "    # Get sample info \n",
    "    sample_info = []\n",
    "    index = df_sample_info.loc[df_sample_info['Sample_No'] == sample].index.item()\n",
    "    for c in columns_for_info_label:\n",
    "        sample_info.append(df_sample_info.at[index, c])\n",
    "    sample_info = ' | '.join(sample_info)\n",
    "\n",
    "    # Total read counts for this sample library\n",
    "    f = open(f'{sample_directory_path}/{total_reads_fname}', \"r\")\n",
    "    library_total_reads = int(f.read()) \n",
    "    f.close()\n",
    "\n",
    "    # Extract out the signature positions (discard those non-signature positions)\n",
    "    df = pd.read_csv(f'{sample_directory_path}/{base_freq_fname}', sep='\\t')\n",
    "    df = df[['ref', 'coord', 'depth', 'A', 'T', 'C', 'G', 'N', 'Del']]\n",
    "    df = df.loc[df['ref'] == SARS2ref]\n",
    "    df = df.loc[df['coord'].isin(coordinates)]\n",
    "    df = df.rename(columns={'Del':'-'}).drop(['ref'], axis=1).set_index('coord')\n",
    "    # if sequencing depth not available for certain signatures, replace with zero   \n",
    "    if len([x for x in coordinates if not x in list(df.index)]) >0:\n",
    "        for p in [x for x in coordinates if not x in list(df.index)]:\n",
    "            df.loc[p] = [0, 0, 0 ,0, 0, 0, 0]\n",
    "\n",
    "    \n",
    "    # Append into the dictionary with the sample No. as key\n",
    "    dict_samples[sample] = [df, library_total_reads, sample_info]\n",
    "    \n",
    "print(dict_samples.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ALLELE FREQUENCY : read counts of variant/ depth\n",
    "\n",
    "# Build a dataframe with signatures as column names and sample numbers as row index\n",
    "df_af = pd.DataFrame(columns = first_lineage_sig + second_lineage_sig, index = list(dict_samples.keys()), dtype=float)\n",
    "\n",
    "# Calculate the allele frequency of the signature variant\n",
    "'''\n",
    "for instance, A21851G(S:K97E) is a signature of BA.2.3.7, \n",
    "the allele frequency will be calculated by (count of reads with 'G'/ depth) at position 21851\n",
    "'''\n",
    "for sample, data in dict_samples.items():\n",
    "    for sig in first_lineage_sig + second_lineage_sig:\n",
    "        coord = int(sig[1:-1])        \n",
    "        alt_base = sig[-1]\n",
    "        # Allele frequency\n",
    "        if data[0].at[coord, 'depth'] == 0:\n",
    "            sig_af = -1\n",
    "        else:\n",
    "            sig_af = data[0].at[coord, alt_base] / data[0].at[coord, 'depth']\n",
    "        # write the value into dataframe\n",
    "        df_af.at[sample, sig] = sig_af\n",
    "\n",
    "df_af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SEQUENCING DEPTH in RPM (Reads per million): \n",
    "# Coverage at a base position / total number of mapped reads in this sample library\n",
    "\n",
    "# Build a dataframe with signatures as column names and sample numbers as row index\n",
    "df_rpm_depth = pd.DataFrame(columns = first_lineage_sig + second_lineage_sig, index = list(dict_samples.keys()), dtype=float)\n",
    "\n",
    "for sig in first_lineage_sig + second_lineage_sig:\n",
    "    coord = sig[1:-1]\n",
    "    df_rpm_depth.rename(columns = {sig: coord}, inplace = True)\n",
    "\n",
    "# Calculate the RPM at signature base, which reflecting the normalized sequencing depth at this position \n",
    "'''\n",
    "for instance, A21851G(S:K97E) is a signature of BA.2.3.7, \n",
    "the depth in RPM will be calculated by (10^6 * count of reads (depth) at 21851 / library total read counts)\n",
    "'''\n",
    "for sample, data in dict_samples.items():\n",
    "    for coord in list(df_rpm_depth.columns):\n",
    "        # Depth\n",
    "        sig_rpm_depth = (pow(10, 6) * data[0].at[int(coord), 'depth']) / data[1]\n",
    "        # write the value into dataframe\n",
    "        df_rpm_depth.at[sample, coord] = sig_rpm_depth\n",
    "\n",
    "df_rpm_depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SAMPLE INFO (Metadata) for the right y-axis label\n",
    "sample_info_list = []\n",
    "for sample, data in dict_samples.items():    \n",
    "    sample_info_list.append(data[2])\n",
    "sample_info_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLOT: ALLELE FREQUENCY & DEPTH IN RPM\n",
    "\n",
    "col_n_1stlineage = len(first_lineage_sig)\n",
    "col_n_2ndlineage = len(second_lineage_sig)\n",
    "\n",
    "fig, ax = plt.subplots(2, figsize=fig_size) \n",
    "\n",
    "# =========== ALLELE FREQUENCY ==========\n",
    "values_af = df_af.to_numpy(dtype=float)\n",
    "\n",
    "sns.heatmap(values_af, cmap=\"RdYlBu_r\", vmin=0, vmax=1, center = 0.5, square=False, \n",
    "            cbar_kws={\"shrink\": 0.9, \"anchor\": (1.5, 0.5)}, ax=ax[0])\n",
    "\n",
    "ax[0].xaxis.tick_top()\n",
    "ax[0].xaxis.set_label_position('top')\n",
    "ax[0].axhline(y = 0, color='k',linewidth = 2)\n",
    "ax[0].axhline(y = values_af.shape[0], color = 'k', linewidth = 2)  \n",
    "ax[0].axvline(x = 0, color = 'k', linewidth = 2)  \n",
    "ax[0].axvline(x = values_af.shape[1], color = 'k', linewidth = 2)\n",
    "ax[0].axvline(x = col_n_1stlineage, color = 'k', linewidth = 3) # seperating line between two lineages\n",
    "\n",
    "sns.heatmap(values_af, xticklabels=df_af.columns, yticklabels=df_af.index,\n",
    "            cmap=plt.get_cmap('binary'), vmin=-2, vmax=0, mask=values_af >= 0, cbar=False, ax=ax[0])\n",
    "\n",
    "ax[0].collections[0].colorbar.set_label(\"Allele Frequency (%)\", fontsize = 14, labelpad=13)\n",
    "ax[0].collections[0].colorbar.ax.tick_params(labelsize=13)\n",
    "ax[0].tick_params(axis='y', length=3, labelrotation=0, pad=8, labelsize = 14)\n",
    "ax[0].tick_params(axis='x', length=3, labelrotation=90, pad=5, labelsize = 14)\n",
    "\n",
    "ax_twin = ax[0].twinx()\n",
    "ax_bbox = ax[0].get_position()\n",
    "plt.subplots_adjust(bottom=ax_bbox.y0, top=ax_bbox.y1)\n",
    "ax_twin.set_ylim(ax[0].get_ylim())\n",
    "ax_twin.set_yticks(ax[0].get_yticks()) \n",
    "ax_twin.set_yticklabels(labels=sample_info_list)\n",
    "ax_twin.tick_params(axis='y', length=0, labelrotation=0, pad=12, labelsize = 14)\n",
    "\n",
    "\n",
    "# ============ DEPTH IN RPM ============\n",
    "values_rpm_depth = df_rpm_depth.to_numpy(dtype=float)\n",
    "\n",
    "# Set limits for color scale according to the values \n",
    "max_value = 4 \n",
    "min_value = -1 \n",
    "#print(np.log10(values_rpm_depth.max()))\n",
    "#print(np.log10(np.extract(values_rpm_depth > 0, values_rpm_depth).min()))\n",
    "\n",
    "# ------ Customed color map ------\n",
    "def create_color(r, g, b):\n",
    "    return [r/256, g/256, b/256]\n",
    " \n",
    "def custom_color_palette():\n",
    "    return LinearSegmentedColormap.from_list(\"\", \n",
    "        [\n",
    "         create_color(215,215,215), #-1, light grey\n",
    "         create_color(255,255,255), #0, white\n",
    "         create_color(255,220,100), #1, light yellow        \n",
    "         create_color(145,220,170), #2, green\n",
    "         create_color(5,82,5),      #3\n",
    "         create_color(0,3,0)        #4\n",
    "        ])\n",
    " \n",
    "cmap = custom_color_palette()\n",
    "# ---------------------------------\n",
    "\n",
    "ax[1].xaxis.tick_top()\n",
    "ax[1].xaxis.set_label_position('top')\n",
    "\n",
    "sns.heatmap(np.log10(values_rpm_depth, where=values_rpm_depth != 0), \n",
    "            cmap=cmap, vmin=min_value, vmax=max_value, square=False, cbar_kws={\"shrink\": 0.9, \"anchor\": (1.5, 0.5)}, ax=ax[1]) \n",
    "\n",
    "ax[1].axhline(y = 0, color='k',linewidth = 3)\n",
    "ax[1].axhline(y = values_rpm_depth.shape[0], color = 'k', linewidth = 3)  \n",
    "ax[1].axvline(x = 0, color = 'k', linewidth = 3)  \n",
    "ax[1].axvline(x = values_rpm_depth.shape[1], color = 'k', linewidth = 3)\n",
    "ax[1].axvline(x = col_n_1stlineage, color = 'k', linewidth = 3)\n",
    "\n",
    "sns.heatmap(values_rpm_depth, xticklabels=df_rpm_depth.columns, yticklabels=df_rpm_depth.index,\n",
    "            cmap=plt.get_cmap('binary_r'), vmin=-1, vmax=1, mask=values_rpm_depth > 0, cbar=False, ax=ax[1])\n",
    "\n",
    "ax[1].collections[0].colorbar.set_label(\"log10(RPM)\", fontsize = 14, labelpad=13)\n",
    "ax[1].collections[0].colorbar.ax.tick_params(labelsize=14)\n",
    "ax[1].tick_params(axis='y', length=3, labelrotation=0, pad=5, labelsize = 14)\n",
    "ax[1].tick_params(axis='x', length=3, labelrotation=90, pad=5, labelsize = 14)\n",
    "\n",
    "\n",
    "# ============ Assign subtitles to subfigures ============\n",
    "ax[0].set_title(subtitle1, x=-0.11, y=1.16, fontweight = 'bold', fontsize = 23)\n",
    "ax[1].set_title(subtitle2, x=-0.11, y=1.16, fontweight = 'bold', fontsize = 23)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(left=0.09)\n",
    "#plt.savefig(f'{working_directory_path}/{patient_id}_AF-depthRPM.png', dpi=600)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
